README.md
[![npm version](https://img.shields.io/npm/v/dnsever-ddns-updater.svg?style=flat-square)](https://www.npmjs.com/package/dnsever-ddns-updater)
[![Build Status](https://travis-ci.org/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater.svg?branch=master)](https://travis-ci.org/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater)

# ddns-update-for-dnsever
Nodejs DDNS Updater, for [DNSEver.com](https://DNSEver.com) ( DNS Service in Korea ).
(DDNS: Dynamic Domain Name Service)

## pre install
프로세스 매니저인 [pm2](https://www.npmjs.com/package/pm2) 환경에서 실행됩니다.
pm2 가 설치되어 있지 않으면 아래의 명령어로 설치합니다.
```
sudo npm install -g pm2
```

## install
```
sudo npm install -g dnsever-ddns-updater
```


## Usage

1. install and show settings
![](https://raw.githubusercontent.com/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater/master/screenshot/01_install_first_run.png)

2. Set your id and auth_code, and check your domain in DNSEver.com
![](https://raw.githubusercontent.com/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater/master/screenshot/02_setting_id_auth_code.png)

3. Run process, wait 5sec, then 'Updater' will try to update your DDNS for your all domain
![](https://raw.githubusercontent.com/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater/master/screenshot/03_start.png)

4. Check new updated DDNS. Without exception, your all domain will be updated as current public_ip
![](https://raw.githubusercontent.com/b6pzeusbc54tvhw5jgpyw8pwz2x6gs/dnsever-ddns-updater/master/screenshot/04_result.png)

5. Every 90sec, 'Updater' compare to public ip and DNSEver.com setting. If update needed, try to update.

6. For more information use help command 
```
dnsever-ddns-updater --help
```



CHANGELOG.md
## 2.3.0
- add .travis.yml and start to use travis-ci
- add some jasmine test
- add jasmine, eslint to devDependencies
- chnage command line color module to colors
- remove unused dependencies json-format, json-beautify


spec/test.js
/* global describe, it, expect */

var core = require('../core');

describe("A suite", function() {

	it("just jasmine describe test.", function() {
		expect(true).toBe(true);
	});

	it("set env, get env", function() {

		var key = 'hello';
		var val = 'world';
		core.setEnv( key, val );
		expect( core.getEnv( key ) === val ).toBe( true );
	});
});


ddnsUpdate.js
var logger = require('tracer').colorConsole();
var request = require( 'request-promise' );
var invariant = require( 'invariant' );
var _ = require( 'underscore');
var xpath = require('xpath'),
	dom = require('xmldom').DOMParser;

var GET_HOST_URL = 'http://dyna.dnsever.com/gethost.php';
var GET_IP_URL = 'http://dyna.dnsever.com/getip.php';
var UPDATE_URL = "http://dyna.dnsever.com/update.php";

function _transformXML( body ) {
	return new dom().parseFromString( body );
}

var realIpAddr = '';
var today = -1;
function _update( id, code, targetHostNameToIp ) {

	var auth = new Buffer( id+':'+code ).toString('base64');
	var	header = {
		'Authorization': 'Basic ' + auth,
		'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36',
	};

	var newDay = (new Date()).getDay();
	if( today !== newDay ) {
		// 적어도 하루에 한번씩은 log를 남긴다
		logger.info( 'Today: ', (new Date()).toLocaleDateString() );
		today = newDay;
	}

	request.get( GET_IP_URL ).then(function( body ) {

		//logger.debug( body );

		if( realIpAddr !== body ) {
			logger.info( 'Public IP address have been changed to ['+body+'] from ['+realIpAddr+']');
			realIpAddr = body;
		}

		//logger.debug( 'req url: ' + GET_HOST_URL );
		//logger.debug( 'headers: ' + JSON.stringify( header ));

		return request.get( GET_HOST_URL, {
			headers: header,
			transform: _transformXML
		});

	}).then( function( doc ) {

		logger.debug( 'gethost res:\n' + doc.toString() );

		var resHostNameList = xpath.select( "//host/@name", doc );
		var resHostIpList = xpath.select( "//host/@ip", doc );

		var qs = {};
		resHostNameList.forEach( function( name, i ) {

			var targetIp = targetHostNameToIp[ name.value ];
			if( targetIp === "IGNORE" ) return;

			if( ! targetIp ) targetIp = realIpAddr;

			if( resHostIpList[i].value !== targetIp ) {
				// 현재 셋팅된 ip와 다를 경우 ddns update 요청을 날릴것이다
				qs[ 'host['+name.value+']' ] = targetIp;
			}
		});

		if( _.isEmpty( qs )) {
			return;
		} else {
			return request.get( UPDATE_URL, {
				headers: header, transform: _transformXML, qs: qs
			});
		}

	}).then( function( doc ) {

		if( ! doc ) return;

		logger.debug( 'Update result:\n' + doc.toString() );
	})
	.catch( function( err ) {
		logger.error( err );
	});
}

function start( id, code, targetHostNameToIp ) {

	setTimeout( function() {

		id = id || process.env.DNSEVER_ID;
		code = code || process.env.DNSEVER_DDNS_SECRET_CODE;
		targetHostNameToIp = targetHostNameToIp || JSON.parse( process.env.DNSEVER_TARGET_HOST_NAME_TO_IP );

		invariant( id, 'No DNSEVER_ID');
		invariant( code, 'No DNSEVER_DDNS_SECRET_CODE');
		invariant( ! _.isEmpty( targetHostNameToIp ), 'Empty targetHostNameToIp');
		invariant( typeof targetHostNameToIp, 'targetHostNameToIp must be "object" type');

		var bindedUpdate = _update.bind( null, id, code, targetHostNameToIp );
		bindedUpdate();

		setInterval( bindedUpdate, 1000 * 90 );

	}, 1000 * 10 );
};

var autoRun = process.env.DNSEVER_AUTO_RUN || "false";
if( JSON.parse( autoRun )) {
	logger.info( require('./package.json').name + ' is run' );
	start();
}

module.exports = {
	start: start
};


ddnsUpdate_old.js
#!/usr/bin/env node
/*jshint devel:true, node:true, undef:true, unused:strict*/

var logger = require('tracer').colorConsole();
var nodefn = require( 'when/node' );
var requestify = require( 'requestify' );
var path = require( 'path' );
var fs = require( 'fs' );
var xpath = require('xpath'),
	dom = require('xmldom').DOMParser;

Error.stackTraceLimit = 100;

var passwd = new Buffer("userid:secretcode").toString('base64');
var	header = { 'Authorization': 'Basic ' + passwd };
var realIpAddr = '';
var today = -1;
var isErrorInPrevTry = false;

var logFileName = 'ddnsUpdate.log';
var logDir = path.join( process.env.HOME, 'uLog' );

nodefn.call( fs.mkdir, logDir )
.then( function() {
	writeLog( 'Has Made the log directory: ' + logdir );
})
.catch( function( err ) {
	writeLog( JSON.stringify( err ));
})
.done( function() {
	setTimeout( update, 10000 );
	setInterval( update, 90000 );
});

function writeLog( str ) {
	console.log( str );
	fs.appendFile( path.join( logDir, logFileName ), str );
}

function update() {

	var isFirstLog = true;

	function log( str ) {
		if( isFirstLog ) {
			isFirstLog = false;
			writeLog( '\n\n------------------- ' + new Date() + ' -------------------\n' );
		}
		writeLog( str );
	}

	var newDay = (new Date()).getDay();
	if( today !== newDay || isErrorInPrevTry ) {
		log( 'I\'m alive\n' );
		today = newDay;
		isErrorInPrevTry = false;
	}

	requestify.get('http://dyna.dnsever.com/getip.php')
	.then(function( res ) {

		res.getBody();
		if( realIpAddr !== res.body ) {
			log( 'Changed IP Address to [' + res.body + '] from [' + realIpAddr + ']');
		}
		realIpAddr = res.body;
		return requestify.get( 'http://dyna.dnsever.com/gethost.php', { headers: header });
	})
	.then( function( res ) {

		res.getBody();

		var doc = new dom().parseFromString( res.body );
		var names = xpath.select( "//host/@name", doc );
		var url = "http://dyna.dnsever.com/update.php?";
		
		names.forEach( function( name ) {
			if(['utopos.me','www.utopos.me'].indexOf( name.value ) < 0 ) return;
			url += 'host['+name.value+']='+realIpAddr+'&';
		});

		// url이 길어지면 post 로 하라는데 잘 안됨.
		return requestify.get( url, { headers: header });
	})
	.then( function( res ) {

		res.getBody();
		var doc = new dom().parseFromString( res.body );
		var msgs = xpath.select( "//host/@msg", doc );

		var isAlreadyUpdated = msgs.every( function( msg ) {
			return msg.value === 'Already Updated';
		});

		if( msgs.length < 2 || !isAlreadyUpdated ) {
			log( res.body );
		}
	})
	.catch( function( err ) {
		log( JSON.stringify( err ));
		isErrorInPrevTry = true;
	});
}


core.js
'use strict';

var request = require( 'request-promise' );
var when = require( 'when' );
var fs = require( 'fs' );
var path = require( 'path' );
var _ = require( 'underscore');

var log4js = require("log4js");
var log4js_extend = require("log4js-extend");
log4js_extend(log4js, { path: __dirname, format: "at @name (@file:@line)" });
var logger = log4js.getLogger("core");
 
var xpath = require('xpath'),
	dom = require('xmldom').DOMParser;

var GET_HOST_URL = 'http://dyna.dnsever.com/gethost.php';
var GET_IP_URL = 'http://dyna.dnsever.com/getip.php';
var UPDATE_URL = "http://dyna.dnsever.com/update.php";
var IP_REGEXP = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?).(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?).(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?).(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;


//
//init
//
var ROOT_PATH = path.resolve(process.env.HOME, '.dnsever-ddns-updater');
var LOG_PATH = path.resolve( ROOT_PATH, 'logs');
//var PROCESS_FILE_PATH = path.resolve( ROOT_PATH, 'process.json' );
var ENV_PATH = path.resolve( ROOT_PATH, 'config.json' );
var env = (function () {

	if ( fs.existsSync( ENV_PATH )) {

		var contents = fs.readFileSync( ENV_PATH, 'utf8' );
		return JSON.parse( contents );

	} else {

		return { id: '',  auth_code: '', exceptionMap: {} };
	}
}());
(function() {
  if (!fs.existsSync( ROOT_PATH )) {
    fs.mkdirSync( ROOT_PATH );
    fs.mkdirSync( LOG_PATH );
  }

  if (process.stdout._handle && process.stdout._handle.setBlocking)
    process.stdout._handle.setBlocking(true);
}());

function _transformXML( body ) {
	return new dom().parseFromString( body );
}

var realIpAddr = '';
var today = -1;
function _tick() {

	var newDay = (new Date()).getDay();
	if( today !== newDay ) {
		// 적어도 하루에 한번씩은 log를 남긴다
		logger.info( 'Today: ', (new Date()).toLocaleDateString() );
		today = newDay;
	}

	getPublicIp().then( function( ip ) {

		if( realIpAddr !== ip ) {
			logger.info('Public IP address have been changed to ['+ip+'] from ['+realIpAddr+']');
			realIpAddr = ip;
		}

		return getCurrentDDNSSetting();

	}).then( function( ddnsSettings ) {

		logger.debug( 'currentDDNSSetting: \n', ddnsSettings );

		var qs = {};
		_.each( ddnsSettings, function( ip, host ) {

			if( env.exceptionMap[ host ] === 'IGNORE' ) return;

			var targetIp;
			if( IP_REGEXP.test( env.exceptionMap[ host ] )) {
				targetIp = env.exceptionMap[ host ];
			} else {
				targetIp = realIpAddr;
			}

			if( ip === targetIp ) return;

			qs[ 'host['+host+']' ] = targetIp;
		});


		if( _.isEmpty( qs )) {
			return;
		} else {
			return request.get( UPDATE_URL, {
				headers: getHeader(), transform: _transformXML, qs: qs
			});
		}

	}).then( function( doc ) {

		if( ! doc ) return;
		logger.debug('Update result:\n' + doc.toString() );
	})
	.catch( function( err ) {
		logger.error( err );
	});
}

function start() {

	var id = env.id;
	var auth_code = env.auth_code;

	setTimeout( function() {

		if( ! id ) {
			logger.error("dnsEver.com 'id' must be set, please command below");
			logger.error('$ dnsever-ddns-updater id your_id');
			return;
		} else if( ! auth_code ) {
			logger.error("dnsEver.com 'auth_code' must be set, please command below");
			logger.error('$ dnsever-ddns-updater auth_code your_auth_code');
		}

		_tick();
		setInterval( _tick, 1000 * 90 );

	}, 1000 * 5 );
};

var autoRun = process.env.DNSEVER_AUTO_RUN || "false";
if( JSON.parse( autoRun )) {
	logger.info( require('./package.json').name + ' is run' );
	start();
}

function deleteHost() {

}

function getHeader() {

	var auth = new Buffer( env.id+':'+env.auth_code ).toString('base64');
	var	header = {
		'Authorization': 'Basic ' + auth,
		'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36',
	};
	return header;
}

function getCurrentDDNSSetting() {

	var deferred = when.defer();

	request.get( GET_HOST_URL, { headers: getHeader(), transform: _transformXML })
	.then( function( doc ) {

		//logger.debug( 'req url: ' + GET_HOST_URL );
		//logger.debug( 'headers: ' + JSON.stringify( header ));
		//logger.debug( 'gethost res:\n' + doc.toString() );

		var msg = xpath.select( "//result/@msg", doc );
		var code = xpath.select( "//result/@code", doc );
		msg = msg[0].value;
		code = code[0].value;

		if( code !== "700" ) {
			deferred.resolve({ errorCode: code, errorMsg: msg });
			return;
		}

		var resHostNameList = xpath.select( "//host/@name", doc );
		var resHostIpList = xpath.select( "//host/@ip", doc );

		var currentDDNSSetting = {};
		var name, ip;

		while( true ) {

			name = resHostNameList.pop();
			ip = resHostIpList.pop();

			if( ! name ) break;
			
			currentDDNSSetting[ name.value ] = ip.value;
		}

		deferred.resolve( currentDDNSSetting );
	});

	return deferred.promise;
}

function save() {
	var envStr = JSON.stringify( env, null, 2 );
	fs.writeFileSync( ENV_PATH, envStr );
}

function setEnv( key, val ) {

	env[ key ] = val;
	save();
}

function getEnv( key ) {
	return env[ key ];
}

function removeException( host ) {

	delete env.exceptionMap[ host ];
	save();
}
function addException( host, ip ) {
	
	env.exceptionMap[ host ] = ip;
	save();
}

function getPublicIp() {
	
	var deferred = when.defer();

	request.get( GET_IP_URL ).then(function( ip ) {

		if ( IP_REGEXP.test( ip ) ) {
			deferred.resolve( ip );
		} else {
			deferred.reject({ errorCode: 'NOT_FOUND_PUBLIC_UP' });
		}
	});

	return deferred.promise;
}

function clean() {
	env = { id: '',  auth_code: '', exceptionMap: {} };
	save();
}
function getEnvPath() {
	return ENV_PATH;
}
function getRootPath() {
	return ROOT_PATH;
}


module.exports = {
	clean: clean,
	deleteHost: deleteHost,
	setEnv: setEnv,
	getEnv: getEnv,
	addException: addException,
	removeException: removeException,
	getPublicIp: getPublicIp,
	getCurrentDDNSSetting: getCurrentDDNSSetting,
	start: start,
	getEnvPath: getEnvPath,
	getRootPath: getRootPath
};
